File: backend/Core/Models/Dtos/Database/ExecutionStepDto.cs
using Core.Enums;

namespace Core.Models.Dtos
{
    public class ExecutionStepDto
    {
        public int Id { get; set; }

        // Where it sits in the run
        public int Sequence { get; set; }
        public int? ParentSequence { get; set; }
        public int Depth { get; set; }
        public int? LoopPass { get; set; }

        public string Name { get; set; } = string.Empty;
        public FlowStepTypeEnum FlowStepType { get; set; }
        public StepOutcomeEnum Outcome { get; set; }
        public DateTime StartedOn { get; set; }
        public int DurationMilliseconds { get; set; }

        public int? ResultLocationX { get; set; }
        public int? ResultLocationY { get; set; }
        public int? MatchIndex { get; set; }
        public int? MatchCount { get; set; }

        // What came back
        public string? Value { get; set; }
        public string? Message { get; set; }


        // SYSTEM_COMMAND
        public int? ExitCode { get; set; }
        public string? Error { get; set; }
        public string? Command { get; set; }

        public string? ResultImagePath { get; set; }

        public int ExecutionId { get; set; }
        public int? FlowStepId { get; set; }
    }
}

File: frontend/src/shared/models/database/execution-step-dto.ts
import type { FlowStepTypeEnum } from "@/shared/enums/backend/flow-step-types-enum";
import type { StepOutcomeEnum } from "@/shared/enums/backend/execution/step-outcome-enum";

/**
 * One step, as it ran. Ordered by sequence and indented by depth it reads as a tree, which is
 * exactly how the run panel draws it - no joins, no walking back up.
 */
export interface ExecutionStepDto {
  id: number;

  // Where it sits in the run
  sequence: number;
  parentSequence?: number | null;
  depth: number;
  loopPass?: number | null;

  name: string;
  flowStepType: FlowStepTypeEnum;
  outcome: StepOutcomeEnum;
  startedOn: string;
  durationMilliseconds: number;

  resultLocationX?: number | null;
  resultLocationY?: number | null;
  matchIndex?: number | null;
  matchCount?: number | null;

  // What came back
  value?: string | null;
  message?: string | null;

  // SYSTEM_COMMAND
  exitCode?: number | null;
  error?: string | null;
  command?: string | null;

  resultImagePath?: string | null;

  executionId: number;
  flowStepId?: number | null;
}
